feat: more detailed grammar scopes for template property binding syntax

dannymcgee opened the pull request:
#### Changes:
- Strictly defined characters in the attribute name pattern (fixes #539)
- Reversed the order of `string.quoted.html` and `punctuation.definition.string.begin/end.html` scopes (TM grammar scopes go from less specific on the left to more specific on the right)

#### Additions:
- **Angular bindings** given a unique scope (`ng-binding.html`) — this scope covers the entire key/value pair, analogous to the HTML grammar's `meta.attribute.<attribute-name>.html` scope
- **Angular binding attribute names** given a unique scope in addition to `entity.other.attribute-name.html` (`entity.other.ng-binding-name.html`) — this scope covers the entire _key_, including punctuation
- **Punctuation** in Angular binding keys given unique scopes:
    - **Brackets and parens** get `punctuation.definition.ng-binding-name.begin/end.html` scopes
    - **Dots** in the attribute name get `punctuation.accessor.html` scope
- Updated template snapshots and added test cases to cover all the changes made above

dannymcgee commented:
@googlebot I signed it!

ayazhafiz commented on syntaxes/template.json:
```diff
@@ -31,25 +31,41 @@
-    "propertyBinding": {
-      "begin": "(\\[.*\\])(=)(\"|')",
+    "ngBinding": {
+      "begin": "(([\\[\\(]{1,2})(?:\\s*)(@?[-_a-zA-Z0-9.$]*)(?:\\s*)([\\]\\)]{1,2}))(=)([\"'])",
```
I would rather keep three separate rules for property, event, and two-way bindings, because a merged pattern will run into issues of matching mixed patterns, like `[prop)`.

dannymcgee replied:
@ayazhafiz Do you think I should go ahead and split them up in this PR, or keep this PR to just property binding and add event/two-way binding in future PRs? I don't mind doing the work here if you wouldn't object to a bigger diff

ayazhafiz replied:
Let's keep this one to property binding and add the rest in separate PRs ))

dannymcgee replied:
@ayazhafiz This should be taken care of now

ayazhafiz commented on syntaxes/template.json:
```diff
@@ -31,25 +31,41 @@
         },
         "2": {
-          "name": "punctuation.separator.key-value.html"
+          "name": "punctuation.definition.ng-binding-name.begin.html"
```
What is the advantage of this custom scope name?

dannymcgee replied:
@ayazhafiz It provides a hook for theme authors to give a unique token color to the punctuation, or to Angular bindings vs regular HTML attributes, etc. The `entity.other.attribute-name` scope is kept as a fallback so themes that haven't gone out of their way to color Angular bindings won't have missing colors.

For example, this is how my theme handles the test cases in this branch:

![Angular template syntax highlighting](https://i.imgur.com/yoa2uaC.png)

dannymcgee commented:
> Thanks for the PR, Danny. I think we should stick to individual rules for property/event/two-way bindings. Could you add some tests for validating/invalidating property binding names?

Added some additional test cases and removed the ones for event/two-way binding. Let me know if there's anything I missed!

dannymcgee commented:
> LGTM aside just the two comments.
@ayazhafiz Should all be taken care of! I left the one comment unresolved so you could review my explanation, but feel free to resolve it if you're satisfied

ayazhafiz commented on syntaxes/template.json:
```diff
@@ -32,30 +32,60 @@
       "beginCaptures": {
         "1": {
-          "name": "entity.other.attribute-name.html"
+          "name": "entity.other.attribute-name.html entity.other.ng-binding-name.html",
```
What do you think of making the scope name more specific, like `ng-property-binding-name`?

dannymcgee replied:
Yeah, that's a good idea now that this scope isn't generic anymore.

dannymcgee replied:
Updated! 0a254fe6fcd05a2f9fd60473328ed83ddd7d72ef

ayazhafiz commented on syntaxes/template.json:
```diff
@@ -32,30 +32,60 @@
+          "name": "string.quoted.html punctuation.definition.string.end.html"
         }
       },
+      "name": "ng-binding.html",
```
Same as above

dannymcgee replied:
Updated! 0a254fe6fcd05a2f9fd60473328ed83ddd7d72ef

For this I took a cue from the HTML grammar and also added a `meta` tag to the scope (explanation [here](https://macromates.com/manual/en/language_grammars#naming_conventions)). The HTML grammar also inserts the property name into their equivalent scope with a backreference, which is something we can think about doing for the built-in directives like `ngStyle`/`ngClass` later on down the road.
